File: 152-maximum-product-subarray/152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        #dp
        maxP = minP = nums[0]
        result = maxP
        for num in nums[1:]:
            tempMax = max(num, maxP * num, minP * num)
            minP = min(num, maxP * num, minP * num)
            maxP = tempMax
            result = max(result, maxP)
        return result

File: 152-maximum-product-subarray/152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        //two arrays
        int n = nums.length;
        int[] max = new int[n];
        int[] min = new int[n];
        max[0] = nums[0];
        min[0] = nums[0];
        int res = nums[0];
        for(int i = 1; i < n; i++) {
            if(nums[i] > 0) {
                max[i] = Math.max(nums[i], nums[i] * max[i - 1]);
                min[i] = Math.min(nums[i], nums[i] * min[i - 1]);
            } else {
                max[i] = Math.max(nums[i], nums[i] * min[i - 1]);
                min[i] = Math.min(nums[i], nums[i] * max[i - 1]);
            }
            res = Math.max(res, max[i]);
        }
        return res;
    }
}